evaluation/tasks/auto_task.py

```diff
@@ -1,25 +1,68 @@
 from abc import ABC, abstractmethod
+from typing import Dict
 import os
 
-from evaluation.utils.io import save_json
+from transformers import AutoTokenizer
+
+from evaluation.utils.io import save_json, load_json
+from evaluation.models import load_model
 
 
 class AutoTask(ABC):
-    def __init__(self, tokenizer, model, device):
-        self.tokenizer = tokenizer
+    def __init__(
+        self, 
+        device, 
+        is_english_only: bool, 
+        model=None, 
+        tokenizer=None,
+        model_name_or_path="",  
+        tokenizer_name="",
+    ):
+        assert model or model_name_or_path, "Expected either `model` or `model_name_or_path`"
+        assert (
+            tokenizer or tokenizer_name or model_name_or_path
+        ), "Expected either `tokenizer` or `model_name_or_path` or `tokenizer_name`"
+        if model is None:
+            model = load_model(model_name_or_path).to(device)
         self.model = model
+        if tokenizer is None:
+            tokenizer = AutoTokenizer.from_pretrained(tokenizer_name or model_name_or_path)
+        self.tokenizer = tokenizer
```

Comment:
I think accepting both model and model_name_or_path (and vice versa for tokenizer) in the class constructor makes it very messy as only one can be used  at a time. The better pattern would be to accept `model` and `tokenizer` in the constructor and offer an alternative to use `model_name_or_path` and `tokenizer_name` in a class method.

Reply:
I would recommend streamlining the classmethods into 2 entries. In this pattern, it's very clear to the user what params need to be passed in without having to worry about which input overrides other inputs. 

```
@classmethod
def from_task_name(cls, task_name: str, **kwargs):
    ...
    ...
    return task(**kwargs)
    
@classmethod
def from_spec(cls, spec: Dict[str, str], **kwargs):
    # spec = { task_name: "tydiqa_secondary",  model_name_or_path: "gpt-2", tokenizer_name: "gpt-2" }
    # we can also choose a different variable name than spec
    # just as long as it indicates that we are constructing the class with multiple strings that need to be mapped
    ...
    ...
    return task(model=model, tokenizer=tokenizer, **kwargs)
```

What do you think?

Reply:
That does indeed look a lot cleaner. I'll go ahead and push a commit with those changes!